charmcraft/main.py

```diff
@@ -0,0 +1,165 @@
+COMMANDS_GROUPS = [
+    ('basic', "basics", [version.VersionCommand, version.CommandExampleError]),
+    ('advanced', "more complex stuff", [version.CommandExampleDebug]),
+]
+
+
+class CustomArgumentParser(argparse.ArgumentParser):
+    """ArgumentParser with grouped commands help."""
+
+    # Flag to indicate action groups that will have custom docs
+    special_group = 'MARKER'
+
+    def __init__(self, **kwargs):
+        self.__commands_groups = kwargs.pop('commands_groups', ())
+        super().__init__(**kwargs)
+
+    def format_help(self):
+        """Produce normal help but showing the commands grouped."""
+        main = False
+        for ag in self._action_groups:
+            if ag.title is self.special_group:
+                self._action_groups.remove(ag)
+                main = True
+                break
+        base = super().format_help()
+        if not main:
+            return base
+
+        # Get the size of the longest name so all helps are columned properly across the groups.
+        longest_name = 0
+        for _, _, cmd_classes in self.__commands_groups:
+            for cmd_class in cmd_classes:
+                longest_name = max(len(cmd_class.name), longest_name)
+
+        extra = ['', 'commands:', '']
+        for group, group_title, cmd_classes in self.__commands_groups:
+            extra.append("    {}:".format(group_title))
+            for cmd_class in sorted(cmd_classes, key=operator.attrgetter('name')):
+                extra.append("        {:{longest}s}   {}".format(
+                    cmd_class.name, cmd_class.help_msg, longest=longest_name))
+            extra.append('')
+
+        return base + '\n'.join(extra)
+
+
+class Dispatcher:
+    """Set up infrastructure and let the needed command run.
+
+    ♪♫"Leeeeeet, the command ruuun"♪♫ https://www.youtube.com/watch?v=cv-0mmVnxPA
+    """
+
+    def __init__(self, sysargs, commands_groups):
+        self.commands = self._load_commands(commands_groups)
+        logger.debug("Commands loaded: %s", self.commands)
+
+        self.main_parser = self._build_argument_parser(commands_groups)
+        self.parsed_args = self.main_parser.parse_args(sysargs)
+        logger.debug("Parsed arguments: %s", self.parsed_args)
+
+    def run(self):
+        """Really run the command."""
+        self._handle_generics()
+
+        if not hasattr(self.parsed_args, '_command'):
+            self.main_parser.print_help()
+            return -1
+
+        command = self.parsed_args._command
+        try:
+            command.run(self.parsed_args)
+        except CommandError as err:
+            return err.retcode
+        return 0
+
+    def _handle_generics(self):
+        """Set up and process generics."""
+        if self.parsed_args.verbose:
+            logsetup.configure('verbose')
+        elif self.parsed_args.quiet:
+            logsetup.configure('quiet')
+
+    def _load_commands(self, commands_groups):
```

Comment:
Well, you registered the commands by classes, and this would instantiate them.

This method is "bringing the commands alive" (and preparing an internal structure while doing so).